add windows setup scripts

StevenJNPearce opened the pull request:
Fixes: #308

antonlerner commented:
Hi Steven, Thank you for your PR!
Please provide a list of prerequisites needed to install on a clean Windows environment, usage instructions and also attach a batch script output that successfully installs and compiles the project 

Thank you

StevenJNPearce commented:
> Hi Steven, Thank you for your PR!
> Please provide a list of prerequisites needed to install on a clean Windows environment, usage instructions and also attach a batch script output that successfully installs and compiles the project
> 
> Thank you

The only prerequisite other than Go is make. With make installed, the process is exactly the same as for linux. Should I add a note to the readme asking users to install make, or add another batch script that calls the setup scripts like the commands in the makefile do?

antonlerner commented:
> > Hi Steven, Thank you for your PR!
> > Please provide a list of prerequisites needed to install on a clean Windows environment, usage instructions and also attach a batch script output that successfully installs and compiles the project
> > Thank you
> 
> The only prerequisite other than Go is make. With make installed, the process is exactly the same as for linux. Should I add a note to the readme asking users to install make, or add another batch script that calls the setup scripts like the commands in the makefile do?

You also use unzip, which should be in the prerequisites. 
we've tried to run an install on a clean windows os and it failed. please provide a screenshot of the full process. If You've encountered any problems and needed to install some additional components please document it

StevenJNPearce commented:
Ok, tomorrow I will spin up a VM with clean windows and see where it fails + screenshot things. It passed on my machine and on appveyor, but I clearly there are some things there not present on a clean install. Log from appveyor:` [00:00:00] Build started
[00:00:00] git clone -q --branch=develop https://github.com/StevenJNPearce/go-spacemesh.git c:\gopath\src\github.com\spacemeshos\go-spacemesh
[00:00:04] git checkout -qf 8c247c5066a17737f889c6a187f9f217467fc6ad
[00:00:05] choco install make
[00:00:06] Chocolatey v0.10.8
[00:00:06] Installing the following packages:
[00:00:06] make
[00:00:06] By installing you accept licenses for the packages.
[00:00:07] 
[00:00:07] Progress: Downloading make 4.2.1... 16%
[00:00:07] Progress: Downloading make 4.2.1... 51%
[00:00:07] Progress: Downloading make 4.2.1... 86%
[00:00:07] Progress: Downloading make 4.2.1... 100%
[00:00:07] 
[00:00:07] make v4.2.1 [Approved]
[00:00:07] make package files install completed. Performing other installation steps.
[00:00:09] Downloading make 64 bit
[00:00:09]   from 'https://sourceforge.net/projects/ezwinports/files/make-4.2.1-without-guile-w32-bin.zip/download'
[00:00:09] 
[00:00:09] Progress: 47% - Saving 160 KB of 336.47 KB
[00:00:09] Progress: 95% - Saving 320 KB of 336.47 KB
[00:00:09] Progress: 100% - Completed download of C:\Users\appveyor\AppData\Local\Temp\1\chocolatey\make\4.2.1\make-4.2.1-without-guile-w32-bin.zip (336.47 KB).
[00:00:09] Download of make-4.2.1-without-guile-w32-bin.zip (336.47 KB) completed.
[00:00:12] Hashes match.
[00:00:12] Extracting C:\Users\appveyor\AppData\Local\Temp\1\chocolatey\make\4.2.1\make-4.2.1-without-guile-w32-bin.zip to C:\ProgramData\chocolatey\lib\make\tools\install...
[00:00:13] C:\ProgramData\chocolatey\lib\make\tools\install
[00:00:14]  ShimGen has successfully created a shim for make.exe
[00:00:14]  The install of make was successful.
[00:00:14]   Software installed to 'C:\ProgramData\chocolatey\lib\make\tools\install'
[00:00:14] 
[00:00:14] Chocolatey installed 1/1 packages. 
[00:00:14]  See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
[00:00:14] set PATH=%PATH%;C:\Users\appveyor\protoc-3.6.1\bin;C:\gopath\bin;
[00:00:14] make all
[00:00:14] setup_env.bat
[00:00:15] ****  Online  ****
[00:00:16]   000000  ...
[00:00:16]   0f5f71
[00:00:16] CertUtil: -URLCache command completed successfully.
[00:00:16] Archive:  C:\Users\appveyor\AppData\Local\Temp\1\protoc-3.6.1-win32.zip
[00:00:16]    creating: C:\Users\appveyor\protoc-3.6.1/include/
[00:00:16]    creating: C:\Users\appveyor\protoc-3.6.1/include/google/
[00:00:16]    creating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/wrappers.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/field_mask.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/api.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/struct.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/descriptor.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/timestamp.proto  
[00:00:16]    creating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/compiler/
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/compiler/plugin.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/empty.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/any.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/source_context.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/type.proto  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/include/google/protobuf/duration.proto  
[00:00:16]    creating: C:\Users\appveyor\protoc-3.6.1/bin/
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/bin/protoc.exe  
[00:00:16]   inflating: C:\Users\appveyor\protoc-3.6.1/readme.txt  
[00:00:16] 
[00:00:16] SUCCESS: Specified value was saved.
[00:00:17] go: finding github.com/grpc-ecosystem/grpc-gateway v1.6.2
[00:00:17] go: finding github.com/golang/snappy v0.0.0-20170215233205-553a64147049
[00:00:17] go: finding github.com/BurntSushi/toml v0.3.1
[00:00:17] go: finding github.com/davecgh/go-xdr v0.0.0-20161123171359-e6a2ba005892
[00:00:17] go: finding github.com/hashicorp/hcl v0.0.0-20180404174102-ef8a98b0bbce
[00:00:17] go: finding github.com/stretchr/testify v0.0.0-20180319223459-c679ae2cc0cb
[00:00:17] go: finding github.com/davecgh/go-spew v1.1.1
[00:00:17] go: finding github.com/google/uuid v0.0.0-20171122162618-4ebdd04351c7
[00:00:17] go: finding github.com/spf13/pflag v1.0.1
[00:00:17] go: finding github.com/golang-collections/go-datastructures v0.0.0-20150211160725-59788d5eb259
[00:00:39] go: finding github.com/spf13/jwalterweatherman v0.0.0-20180109140146-7c0cea34c8ec
[00:00:39] go: finding github.com/seehuhn/mt19937 v0.0.0-20180715112136-cc7708819361
[00:00:39] go: finding github.com/spf13/cast v1.2.0
[00:00:39] go: finding gopkg.in/natefinch/lumberjack.v2 v2.0.0-20170531160350-a96e63847dc3
[00:00:39] go: finding github.com/pmezard/go-difflib v1.0.0
[00:00:39] go: finding github.com/inconshreveable/mousetrap v1.0.0
[00:00:41] go: finding github.com/spf13/viper v0.0.0-20180507071007-15738813a09d
[00:00:41] go: finding golang.org/x/crypto v0.0.0-20171128194009-94eea52f7b74
[00:00:41] go: finding github.com/golang/protobuf v1.2.0
[00:00:41] go: finding github.com/onsi/gomega v1.4.3
[00:00:42] go: finding github.com/spf13/afero v1.1.0
[00:00:42] go: finding github.com/magiconair/properties v1.8.0
[00:00:43] go: finding golang.org/x/net v0.0.0-20180906233101-161cd47e91fd
[00:00:43] go: finding github.com/btcsuite/btcd v0.0.0-20171128150713-2e60448ffcc6
[00:00:43] go: finding golang.org/x/sync v0.0.0-20181108010431-42b317875d0f
[00:00:43] go: finding github.com/pelletier/go-toml v0.0.0-20180323185243-66540cf1fcd2
[00:00:45] go: finding github.com/onsi/ginkgo v1.6.0
[00:00:45] go: finding google.golang.org/genproto v0.0.0-20171123000638-7f0da29060c6
[00:00:45] go: finding github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
[00:00:47] go: finding golang.org/x/text v0.3.0
[00:00:47] go: finding gopkg.in/yaml.v2 v2.2.1
[00:00:47] go: finding github.com/fsnotify/fsnotify v1.4.7
[00:00:47] go: finding github.com/btcsuite/btcutil v0.0.0-20170726183619-501929d3d046
[00:00:48] go: finding gopkg.in/tomb.v1 v1.0.0-20141024135613-dd632973f1e7
[00:00:51] go: finding github.com/onsi/ginkgo v1.7.0
[00:00:51] go: finding golang.org/x/sys v0.0.0-20180909124046-d0be0721c37e
[00:00:51] go: finding golang.org/x/sync v0.0.0-20180314180146-1d60e4601c6f
[00:00:52] go: finding google.golang.org/grpc v0.0.0-20171201233205-cd563b81ec33
[00:00:52] go: finding gopkg.in/fsnotify.v1 v1.4.7
[00:00:52] go: finding github.com/spf13/cobra v0.0.3
[00:00:55] go: finding gopkg.in/yaml.v2 v2.2.2
[00:00:55] go: finding github.com/syndtr/goleveldb v0.0.0-20171214120811-34011bf325bc
[00:00:55] go: finding github.com/hpcloud/tail v1.0.0
[00:00:56] go: finding gopkg.in/op/go-logging.v1 v1.0.0-20160211212156-b2cb9fa56473
[00:00:56] go: finding github.com/gogo/protobuf v0.0.0-20171204084257-41168f6614b7
[00:00:56] go: finding github.com/mitchellh/mapstructure v0.0.0-20180511142126-bb74f1db0675
[00:00:57] go: finding gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405
[00:01:16] go: finding github.com/golang/protobuf/protoc-gen-go latest
[00:01:17] go: downloading github.com/golang/protobuf v1.2.0
[00:01:40] make genproto
[00:01:40] make[1]: Entering directory 'C:/gopath/src/github.com/spacemeshos/go-spacemesh'
[00:01:40] scripts\win\genproto.bat
[00:01:40] Generating protobuf for C:\gopath\src\github.com\spacemeshos\go-spacemesh\api\pb\api.proto
[00:01:40] Generating protobuf for C:\gopath\src\github.com\spacemeshos\go-spacemesh\consensus\pb\messages.proto
[00:01:40] Generating protobuf for C:\gopath\src\github.com\spacemeshos\go-spacemesh\hare\pb\hare.proto
[00:01:41] Generating protobuf for C:\gopath\src\github.com\spacemeshos\go-spacemesh\merkle\pb\node.proto
[00:01:41] Generating protobuf for C:\gopath\src\github.com\spacemeshos\go-spacemesh\p2p\dht\pb\findnode.proto
[00:01:41] Generating protobuf for C:\gopath\src\github.com\spacemeshos\go-spacemesh\p2p\pb\message.proto
[00:01:41] Generating protobuf for C:\gopath\src\github.com\spacemeshos\go-spacemesh\ping\pb\ping.proto
[00:01:41] Generating protobuf for C:\gopath\src\github.com\spacemeshos\go-spacemesh\sync\pb\sync.proto
[00:01:41] make[1]: Leaving directory 'C:/gopath/src/github.com/spacemeshos/go-spacemesh'
[00:01:41] go build -ldflags "-X main.version=0.0.1 -X main.commit=8c247c5066a17737f889c6a187f9f217467fc6ad -X main.branch=HEAD" -o /c/gopath/src/github.com/spacemeshos/go-spacemesh/go-spacemesh
[00:01:41] go: downloading github.com/spf13/cobra v0.0.3
[00:01:41] go: downloading github.com/spf13/viper v0.0.0-20180507071007-15738813a09d
[00:01:41] go: downloading gopkg.in/op/go-logging.v1 v1.0.0-20160211212156-b2cb9fa56473
[00:01:41] go: downloading github.com/btcsuite/btcutil v0.0.0-20170726183619-501929d3d046
[00:01:41] go: downloading github.com/grpc-ecosystem/grpc-gateway v1.6.2
[00:01:41] go: downloading google.golang.org/grpc v0.0.0-20171201233205-cd563b81ec33
[00:01:41] go: downloading golang.org/x/crypto v0.0.0-20171128194009-94eea52f7b74
[00:01:41] go: downloading github.com/spf13/pflag v1.0.1
[00:01:41] go: downloading github.com/google/uuid v0.0.0-20171122162618-4ebdd04351c7
[00:01:41] go: downloading gopkg.in/natefinch/lumberjack.v2 v2.0.0-20170531160350-a96e63847dc3
[00:01:42] go: downloading golang.org/x/net v0.0.0-20180906233101-161cd47e91fd
[00:01:42] go: downloading github.com/hashicorp/hcl v0.0.0-20180404174102-ef8a98b0bbce
[00:01:42] go: downloading github.com/spf13/cast v1.2.0
[00:01:42] go: downloading github.com/spf13/jwalterweatherman v0.0.0-20180109140146-7c0cea34c8ec
[00:01:42] go: downloading github.com/pelletier/go-toml v0.0.0-20180323185243-66540cf1fcd2
[00:01:42] go: downloading github.com/mitchellh/mapstructure v0.0.0-20180511142126-bb74f1db0675
[00:01:42] go: downloading github.com/inconshreveable/mousetrap v1.0.0
[00:01:42] go: downloading github.com/btcsuite/btcd v0.0.0-20171128150713-2e60448ffcc6
[00:01:43] go: downloading gopkg.in/yaml.v2 v2.2.2
[00:01:43] go: downloading github.com/spf13/afero v1.1.0
[00:01:43] go: downloading github.com/fsnotify/fsnotify v1.4.7
[00:01:43] go: downloading github.com/gogo/protobuf v0.0.0-20171204084257-41168f6614b7
[00:01:43] go: downloading github.com/magiconair/properties v1.8.0
[00:01:43] go: downloading google.golang.org/genproto v0.0.0-20171123000638-7f0da29060c6
[00:01:44] go: downloading github.com/davecgh/go-spew v1.1.1
[00:01:44] go: downloading golang.org/x/text v0.3.0
[00:01:46] go: downloading github.com/syndtr/goleveldb v0.0.0-20171214120811-34011bf325bc
[00:02:02] Discovering tests...OK
[00:02:02] Build success`

StevenJNPearce commented:
Got everything working on the clean install VM. Added 'Windows' section in the README.md with windows specific instructions, also removed unzip as a dependancy, now we call the native windows unzip functionality. Screenshots here: https://imgur.com/a/tefCyKD

antonlerner commented:
Hey Steven, thank you for your PR, we will review and get back to you ASAP

antonlerner commented on README.md:
```diff
@@ -123,6 +123,13 @@ docker build -t spacemesh .
+- Git for Windows - after installation remove `C:\Program Files\Git\bin` from System PATH and add `C:\Program Files\Git\cmd` to System PATH
+- Make - after installation add `C:\Program Files (x86)\GnuWin32\bin` to System PATH
+
+You can then run the commands `make install` and `make build` as on unix based systems.
```
please specify the order of the commands, i.e first run make install and then make build

antonlerner commented on setup_env.bat:
```diff
@@ -0,0 +1,8 @@
+@echo off
+call ./scripts/win/install-protobuf.bat
```
getting this error, even after running as administrator
Used the virtual box win10 vm 

setup_env.bat
Access is denied.
Exception calling "ExtractToDirectory" with "2" argument(s): "Could not find file
'C:\Users\User\AppData\Local\Temp\protoc-3.6.1-win32.zip'."
At line:1 char:53
+ ... ileSystem'; [IO.Compression.ZipFile]::ExtractToDirectory('C:\Users\Us ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : FileNotFoundException


SUCCESS: Specified value was saved.
go: finding github.com/golang/protobuf/protoc-gen-go latest

StevenJNPearce replied:
Odd I can't reproduce and didn't require administrator; but will change the download to use something other than certutil - it's a bit of an odd hack I suspect it is the cause.

antonlerner commented on Makefile:
```diff
@@ -18,16 +19,31 @@ all: install build
 .PHONY: genproto
 
 build:
+ifeq ($(OS),Windows_NT)
```
getting this error

make genproto
make[1]: Entering directory `C:/Users/User/go-spacemesh'
scripts\win\genproto.bat
Generating protobuf for C:\Users\User\go-spacemesh\api\pb\api.proto
The system cannot find the path specified.
Generating protobuf for C:\Users\User\go-spacemesh\consensus\pb\messages.proto
The system cannot find the path specified.
Generating protobuf for C:\Users\User\go-spacemesh\hare\pb\hare.proto
The system cannot find the path specified.
Generating protobuf for C:\Users\User\go-spacemesh\merkle\pb\node.proto
The system cannot find the path specified.
Generating protobuf for C:\Users\User\go-spacemesh\p2p\dht\pb\findnode.proto
The system cannot find the path specified.
Generating protobuf for C:\Users\User\go-spacemesh\p2p\pb\message.proto
The system cannot find the path specified.
Generating protobuf for C:\Users\User\go-spacemesh\ping\pb\ping.proto
The system cannot find the path specified.
Generating protobuf for C:\Users\User\go-spacemesh\sync\pb\sync.proto
The system cannot find the path specified.
The system cannot find the path specified.
The system cannot find the path specified.
make[1]: Leaving directory `C:/Users/User/go-spacemesh'
go build -ldflags "-X main.version=0.0.1 -X main.commit=f8d87ac48afd6aa2de24365f163776eda979e282 -X main.branch=develop" -o C:\Users\User\go-spacemesh/go-spacemesh
go: finding github.com/spacemeshos/go-spacemesh/p2p/pb latest
go: finding github.com/spacemeshos/go-spacemesh/p2p/dht/pb latest
go: finding github.com/spacemeshos/go-spacemesh/api/pb latest
go: finding github.com/spacemeshos/go-spacemesh/p2p latest
go: finding github.com/spacemeshos/go-spacemesh/p2p/dht latest
go: finding github.com/spacemeshos/go-spacemesh/api latest
api\grpc_server.go:6:2: unknown import path "github.com/spacemeshos/go-spacemesh/api/pb": cannot find module providing package github.com/spacemeshos/go-spacemesh/api/pb
p2p\dht\findnode.go:9:2: unknown import path "github.com/spacemeshos/go-spacemesh/p2p/dht/pb": cannot find module providing package github.com/spacemeshos/go-spacemesh/p2p/dht/pb
p2p\net\handshake.go:19:2: unknown import path "github.com/spacemeshos/go-spacemesh/p2p/pb": cannot find module providing package github.com/spacemeshos/go-spacemesh/p2p/pb
make: *** [build] Error 1

StevenJNPearce commented:
@antonlerner I've been unable to reproduce these errors. It is working fine for me using both the Windows 10 dev environment in Hyper-V and Windows 10 Pro in VirtualBox.

StevenJNPearce commented:
I've gone ahead and replaced the download method for protoc; it was previously using certutil.exe which is a bit of an odd hack. I suspect this is what was causing the issue for you.

antonlerner commented:
Hi @StevenJNPearce , Thanks for your updated version, I ran it and it seem to compile the file, however an .exe file was not created and spacemeh binary could not be run, could you please look into this?

StevenJNPearce commented:
> Hi @StevenJNPearce , Thanks for your updated version, I ran it and it seem to compile the file, however an .exe file was not created and spacemeh binary could not be run, could you please look into this?

Fixed, but now there is a merge confict, one moment while I address that.

StevenJNPearce commented:
> > Hi @StevenJNPearce , Thanks for your updated version, I ran it and it seem to compile the file, however an .exe file was not created and spacemeh binary could not be run, could you please look into this?
> 
> Fixed, but now there is a merge confict, one moment while I address that.

Done.

antonlerner commented on scripts/win/install-protobuf.bat:
```diff
@@ -0,0 +1,6 @@
+ECHO Fetching protoc...
+powershell.exe -nologo -noprofile -command "& { [Net.ServicePointManager]::SecurityProtocol = 'tls12, tls11, tls'; (new-object System.Net.WebClient).DownloadFile('https://github.com/protocolbuffers/protobuf/releases/download/v3.6.1/protoc-3.6.1-win32.zip','%TEMP%\protoc-3.6.1-win32.zip'); }"
+ECHO Extracting protoc...
+powershell.exe -nologo -noprofile -command "& { Add-Type -A 'System.IO.Compression.FileSystem'; [IO.Compression.ZipFile]::ExtractToDirectory('%TEMP%\protoc-3.6.1-win32.zip', '%USERPROFILE%\protoc-3.6.1'); }"
```
Is powershell integrated in all windows systems by default? if not, it should be included in prerequisites

StevenJNPearce replied:
@antonlerner The last version of Windows not to include it was Vista, which is now over 12 years old with less than 1% market share and stopped receiving critical security updates years ago. Do you think it's reasonable to leave it out from prerequisites to avoid cluttering the README with unnecessary information?

StevenJNPearce committed:
add windows setup scripts
